### PR TITLE
Add support for __gnuc_va_list to ClangImporter

### PR DESCRIPTION
**Pull request by hpux735**
The implementation of va_list is pretty inconsistent across platforms.  On linux-arm __gnuc_va_list is the type used.  This change adds that to ClangImporter.

The change to the test is optional, but it dramatically improves the error message printed to the console when the test fails.  

Finally, the change to the append method of VaListBuilder ensures proper alignment to AAPCS, which is expected by linux-arm as well as watchos-arm.

Related issue: https://github.com/SwiftAndroid/swift/issues/15#issuecomment-182967084 @jrose-apple

### CODE REVIEW

File: test/1_stdlib/VarArgs.swift [hunk 1]
```diff
@@ -12,8 +12,8 @@ typealias CGFloat = Double
 #endif
 
 func my_printf(format: String, _ arguments: CVarArgType...) {
-  withVaList(arguments) {
-    vprintf(format, $0)
+  withVaList(arguments) { (vaList: CVaListPointer) in
+    vprintf(format, vaList)
```
Comment (modocache) [inline test/1_stdlib/VarArgs.swift]: Technically, we don't need to include this change to get the test passing. However, it did made it easier to spot that `__gnuc_va_list` wasn't being mapped properly.

On the one hand, leaving it in will make it easier for other ports--if the `va_list` mappings don't apply they'll immediately see what needs to be added to `MappedTypes.def`. On the other, it isn't strictly necessary.

Comment (hpux735) [inline test/1_stdlib/VarArgs.swift]: Exactly.  Unless there are subtle side effects to leaving it in, I really like the informative error messages that this change enables.

File: stdlib/public/core/VarArgs.swift [hunk 2]
```diff
@@ -290,7 +290,9 @@ final public class VaListBuilder {
     // supported vararg type is greater than the alignment of Int.
     // FIXME: this implementation is not portable because
     // alignof differs from the ABI alignment on some architectures
-#if os(watchOS) && arch(arm)   // FIXME: rdar://21203036 should be arch(armv7k)
+    // FIXME: This is valid for AAPCS calling convention, FreeBSD/arm
+    // may need this too.
+#if (os(watchOS) || os(Linux)) && arch(arm)   // FIXME: rdar://21203036 should be arch(armv7k)
```
Comment (jrose-apple) [inline stdlib/public/core/VarArgs.swift]: Let's just do the right thing: `!(os(iOS) && arch(arm))`

File: test/1_stdlib/VarArgs.swift (hunk 1 above)
Comment (jrose-apple) [inline test/1_stdlib/VarArgs.swift]: Seems reasonable to me.

File: stdlib/public/core/VarArgs.swift (hunk 2 above)
Comment (tienex) [inline stdlib/public/core/VarArgs.swift]: Are other 32-bits architectures susceptible of this? Wouldn't be best to be !os(iOS) && arch(arm)? Does this enable the tvOS target to use it or iOS includes tvOS?

Comment (jrose-apple) [inline stdlib/public/core/VarArgs.swift]: iOS does not include tvOS, but there's no 32-bit tvOS. I think we should make normal AAPCS the default and opt out for Apple platforms rather than the other way around.

I guess you're right about not touching non-ARM platforms for now.

Comment (tienex) [inline stdlib/public/core/VarArgs.swift]: Cool, thanks!

Comment (hpux735) [inline stdlib/public/core/VarArgs.swift]: @jrose-apple I'm about to push this fix, but I want to ask whether you want to keep the fixme (referencing the rdar) about armv7k.  It doesn't seem like a v7k issue.

Comment (jrose-apple) [inline stdlib/public/core/VarArgs.swift]: Nah, it's more "any other hypothetical platforms using the same ABI as watchOS", i.e. if Apple did another 32-bit ARM platform, it's likely to be more like watchOS than iOS.

Comment (jrose-apple) [inline stdlib/public/core/VarArgs.swift]: Then again there could be an ABI-breaking change to 32-bit iOS to use the same ABI as watchOS, and then we'd call that "armv7k" as well, so I suppose it's not completely irrelevant.

File: stdlib/public/core/VarArgs.swift [hunk 3]
```diff
@@ -290,7 +290,8 @@ final public class VaListBuilder {
     // supported vararg type is greater than the alignment of Int.
     // FIXME: this implementation is not portable because
     // alignof differs from the ABI alignment on some architectures
-#if os(watchOS) && arch(arm)   // FIXME: rdar://21203036 should be arch(armv7k)
+    // All ARM platforms with the exception of iOS use the AAPCS conventions
+#if !os(iOS) && arch(arm)   // FIXME: rdar://21203036 should be arch(armv7k)
```
Comment (hpux735) [inline stdlib/public/core/VarArgs.swift]: @jrose-apple git hub again lost all the comments in the update, but here's the corrected line.  I left the FIXME because that seemed to be your (slight?) preference.

Comment (jrose-apple) [inline stdlib/public/core/VarArgs.swift]: Yeah, although I'd probably reverse the condition and maybe rephrase the whole comment to explain things better. That shouldn't block the merge, though.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
…although this should have no effect on any of the platforms we're actually testing in that.

**Comment by hpux735** [comment]
That's the hope, of course, but it's worth checking :)

**Comment by jrose-apple** [comment]
Sorry I didn't ask sooner: can you update the commit message to note the resolution of the float issue? (Which may also be an Int64 issue.)

**Comment by hpux735** [comment]
No problem.  I reworded the comment, reversed the order of the comparison (as mentioned earlier) and revised the commit comment.  Thanks @jrose-apple !

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by hpux735** [comment]
Sounds good, @jrose-apple .  I used your comment. :)

**Comment by gparker42** [comment]
watchOS armv7k uses a variant of AAPCS where the maximum alignment is 16 bytes. I think the code here is fine unless and until we need to pass vector parameters through a vararg interface.
